### PR TITLE
Initial integration with FairScale Pipe Module for model partitioning/gradient checkpointing

### LINKED ISSUE #4322: Add model sharding/gradient checkpointing from FairScale
## 🚀 Feature

Recently in [this](https://ai.facebook.com/research/publications/beyond-english-centric-multilingual-machine-translation) paper from Facebook Research they discuss making huge memory savings using checkpointing/pipe parallel and optimizer state checkpointing to train their transformer model.

We've already got integration with #4178 for optimizer state sharding (deepspeed) and with sharded data parallel, but might be worth working with the fairscale team to integrate gradient checkpointing and eventually model pipe sharding (I know Ben is working on this).

cc @williamFalcon @blefaudeux

Once the PR is merged for the initial integration, I'll outline a proposal on how we can introduce these features as configuration options to the fairscale accelerator.

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #4322.

Relates to #4178 but I want to keep separate the FairScale ZeRO + ShardedDDP integration and the Pipe + Checkpointing integration since they're decoupled (for now) in fairscale.

Pipe allows a sequential model to be split onto separate GPUs. It comes with its own hyper-parameters and because it's tied to a `torch.nn.sequential` it needs closer integration from the users perspective. 

Still unsure on the API, but want to throw something out for us to discuss.

All the feedback welcomed :)

```python
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = BoringModel()
    # model.layers is a sequential module that needs to be manually wrapped 
    model.layers = LightningPipeModule(
        model.layers,
        layer_partitions=[1, 1], # Puts 1 layer on each GPU
        microbatches=8 # Bubble partitioning under the hood for device utilization
    )
    accelerator = PipeAccelerator(model.layers, cluster_environment=TorchElasticEnvironment())

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        gpus=2,
        accelerator=accelerator
    )
    trainer.fit(model, train_data, val_data)
```

## TODO

- [x] Fix backward pass, such that non last layer rank GPUs call `backward_helper` rather than calling the loss
- [ ] Test gradient checkpointing through Pipe
- [x] Fix evaluation with Pipe module (see https://github.com/facebookresearch/fairscale/issues/173)

cc @ananthsub @williamFalcon 

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 1]
```diff
@@ -186,6 +187,9 @@ def select_accelerator(self):
         if self.accelerator is not None and isinstance(self.accelerator, Accelerator):
             self.accelerator.trainer = self.trainer
             self.accelerator.ddp_plugin = self.trainer.plugin_connector.ddp_plugin
+            if type(self.accelerator) == PipeAccelerator:
```
Comment (SeanNaren) [inline pytorch_lightning/accelerators/accelerator_connector.py:190]: Hacky as hell for now...

File: pytorch_lightning/accelerators/pipe_accelerator.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            balance=self.layer_partitions,
+            chunks=self.microbatches,
+            style=PipelineStyle.MultiProcess,
+            worker_map=get_worker_map()
+        )
+
+    def forward(self, *args, **kwargs):
+        x = self.module(*args, **kwargs)
+        if not self.training:
+            x = microbatch.gather(x)
```
Comment (SeanNaren) [inline pytorch_lightning/accelerators/pipe_accelerator.py:45]: This doesn't seem correct, there needs to be some form of sync happening (hopefully answered by https://github.com/facebookresearch/fairscale/issues/173)

**Comment by SeanNaren** [comment]
So the code works, allowing the user to split the model across GPUs in a SPSD fashion (I've just made that hardcoded, not sure we want to worry about SPMD support). 

Buuut imo it involves too much of the users input, particulary in their train/val/test step logic:

```python
    def training_step(self, batch, batch_idx):
        output = self(batch)
        if self.trainer.accelerator_backend.final_stage:
            loss = self.loss(batch, output)
            self.log('loss', loss)
            print(loss)
            return {"loss": loss}
        else:
            self.trainer.accelerator_backend.sync_gradients(output)

    def validation_step(self, batch, batch_idx):
        output = self(batch)
        if self.trainer.accelerator_backend.final_stage:
            loss = self.loss(batch, output)
            self.log('x', loss)

    def test_step(self, batch, batch_idx):
        output = self.layers(batch)
        if self.trainer.accelerator_backend.final_stage:
            loss = self.loss(batch, output)
            self.log('y', loss)
```

This check is required because some GPUs are just intermediates (they only contain the portion of the model that the activations are passed downstream to other processes, no loss calculated). This happens to include GPU 0 so right now logging is a mess.

Theres' a neat refactor from @froody which will allow control from just one process, which should make things much cleaner and potentially remove the manual final stage check within the step functions (https://github.com/facebookresearch/fairscale/issues/173#issuecomment-719842863)

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator_connector.py:190]: Maybe worth to refactor/cleanup this part. It would look smoother.

File: pytorch_lightning/accelerators/pipe_accelerator.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    def init_pipe(self):
+        self.module = Pipe(
+            module=self.module,
+            balance=self.layer_partitions,
+            chunks=self.microbatches,
+            style=PipelineStyle.MultiProcess,
+            input_device=torch.cuda.current_device(),
+            worker_map=get_worker_map(),
+            checkpoint='never',
```
Comment (tchaton) [inline pytorch_lightning/accelerators/pipe_accelerator.py:38]: What is `checkpoint` never ?

File: pytorch_lightning/accelerators/pipe_accelerator.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        init_method = f"tcp://{os.environ['MASTER_ADDR']}:{os.environ['MASTER_PORT']}"
+
+        rpc.init_rpc(
+            f"Test{global_rank}",
+            rank=global_rank,
+            world_size=world_size,
+            backend=rpc.BackendType.TENSORPIPE,
+            rpc_backend_options=rpc.TensorPipeRpcBackendOptions(init_method=init_method),
+        )
+        mpu.initialize_model_parallel(model_parallel_size_=1, pipeline_length=len(self.pipe_module.layer_partitions))
```
Comment (tchaton) [inline pytorch_lightning/accelerators/pipe_accelerator.py:81]: May the user want to modify those parameters ?

**Review by tchaton — COMMENTED** [review]
Awesome addition ! I wonder if we could block function call using `self.trainer.accelerator_backend.final_stage` in self.log. However, I guess `sync_gradients` can't be moved out as it needs output.

**Comment by SeanNaren** [comment]
Thanks for your comments @tchaton :)

I had a question around the API, since I think this is eventually going to tie to a few other important components (like standalone gradient checkpointing, and potentially even the shardedDDP stuff).

In the current API I've wrapped the sequential module in a lightning module, and defined a custom accelerator in the code:

```python
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = BoringModel()
    # model.layers is a sequential module that needs to be manually wrapped 
    model.layers = LightningPipeModule(
        model.layers,
        layer_partitions=[1, 1], # Puts 1 layer on each GPU
        microbatches=8 # Bubble partitioning under the hood for device utilization
    )
    accelerator = PipeAccelerator(model.layers, cluster_environment=TorchElasticEnvironment())

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        gpus=2,
        accelerator=accelerator
    )
    trainer.fit(model, train_data, val_data)
```

This is meh and I'd prefer to so something like:

```python
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = BoringModel()
    # model.layers is a sequential module that needs to be manually wrapped 
    model.layers = LightningPipeModule(
        model.layers,
        layer_partitions=[1, 1], # Puts 1 layer on each GPU
        microbatches=8 # Bubble partitioning under the hood for device utilization
    )

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        gpus=2,
        accelerator='ddp_pipe' # Skip initializing the accelerator beforehand!
    )
    trainer.fit(model, train_data, val_data)
```

Is there a way I could 'register' the pipe module for the accelerator to pick up automatically?

File: pytorch_lightning/accelerators/pipe_accelerator.py (hunk 3 above)
Comment (froody) [reply pytorch_lightning/accelerators/pipe_accelerator.py]: The `checkpoint` argument refers to activation checkpointing, i.e. not saving grads on the initial forward pass, and then re-running the forward pass with grads enabled during the backward pass.

Comment (SeanNaren) [reply pytorch_lightning/accelerators/pipe_accelerator.py]: Will be exposed eventually of course :)
